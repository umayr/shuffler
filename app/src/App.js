import { useState } from 'react'
import {
    Grommet,
    Main,
    TextInput,
    Box,
    Button,
    List,
    DropButton,
    Keyboard,
} from 'grommet'
import { grommet } from 'grommet/themes'
import {
    Add as AddIcon,
    Group as GroupIcon,
    ClearOption as ClearIcon,
    Sign as PickIcon,
    Down as DownIcon,
    Cycle as ShuffleIcon,
    TreeOption as SeedIcon,
    Keyboard as ShortcutIcon,
} from 'grommet-icons'

import { Group, Pick, Help } from './Modals'

import useMode from './useMode'
import useShuffle from './useShuffle'

const KEY_MAPPING = [
    {
        key: 'ctrl + h or ?',
        text: 'Shows the help screen',
    },
    {
        key: 'ctrl + g',
        text: 'Opens the screen to group items together',
    },
    {
        key: 'ctrl + p',
        text: 'Opens the screen to pick items',
    },
    {
        key: 'ctrl + s',
        text: 'Shuffles items',
    },
    {
        key: 'ctrl + c',
        text: 'Clears all previous data',
    },
    {
        key: 'ctrl + d',
        text: 'Toggles between light and dark mode',
    },
]

function App() {
    const [value, setValue] = useState('')

    const [showPick, setShowPick] = useState(false)
    const [showGroup, setShowGroup] = useState(false)
    const [showHelp, setShowHelp] = useState(false)

    const [mode, toggleMode] = useMode()
    const { list, push, shuffle, clear, seed, isEmpty } = useShuffle()

    const handleAdd = () => {
        push(value)
        setValue('')
    }

    const handleKey = event => {
        if (!event.ctrlKey && event.key !== '?') {
            return
        }

        const key = event.key.toLowerCase()

        if (!['?', 'h', 'd'].includes(key) && isEmpty) {
            // prevent shuffle actions in case of the source being empty
            return
        }

        switch (key) {
            case 'h':
            case '?':
                setShowHelp(true)
                return

            case 'd':
                toggleMode()
                return

            case 'p':
                setShowPick(true)
                return

            case 'g':
                setShowGroup(true)
                return

            case 'c':
                clear()
                return

            case 's':
                shuffle()
                return

            default:
                return
        }
    }

    return (
        <Grommet
            theme={grommet}
            themeMode={mode}
            style={{ minHeight: '100vh' }}
        >
            {showPick && (
                <Pick source={list} onClose={() => setShowPick(false)} />
            )}

            {showGroup && (
                <Group source={list} onClose={() => setShowGroup(false)} />
            )}

            {showHelp && (
                <Help source={KEY_MAPPING} onClose={() => setShowHelp(false)} />
            )}

            <Keyboard onKeyDown={handleKey} target="document">
                <Main pad="large">
                    <Box direction="column" gap="medium">
                        <Keyboard onEnter={handleAdd}>
                            <Box direction="row" gap="medium">
                                <TextInput
                                    placeholder="Anything goes here"
                                    value={value}
                                    onChange={event =>
                                        setValue(event.target.value)
                                    }
                                />

                                <Button
                                    primary
                                    icon={<AddIcon />}
                                    label="Add"
                                    disabled={value.length === 0}
                                    onClick={handleAdd}
                                />
                            </Box>
                        </Keyboard>
                        <List data={list} pad="small" />
                        <Box
                            direction="row-reverse"
                            gap="medium"
                            height="xxsmall"
                        >
                            <DropButton
                                label="More"
                                icon={<DownIcon />}
                                dropAlign={{ top: 'bottom', right: 'right' }}
                                dropContent={
                                    <Box pad="medium" gap="medium">
                                        <Button
                                            primary
                                            icon={<ClearIcon />}
                                            label="Clear"
                                            disabled={isEmpty}
                                            onClick={clear}
                                        />
                                        <Button
                                            primary
                                            icon={<PickIcon />}
                                            label="Pick"
                                            disabled={isEmpty}
                                            onClick={() => setShowPick(true)}
                                        />
                                        <Button
                                            primary
                                            icon={<GroupIcon />}
                                            label="Group"
                                            disabled={isEmpty}
                                            onClick={() => setShowGroup(true)}
                                        />
                                        <Button
                                            primary
                                            icon={<ShortcutIcon />}
                                            label="Shortcuts"
                                            onClick={() => setShowHelp(true)}
                                        />
                                        <TextInput
                                            name="seed"
                                            placeholder="seed"
                                            icon={<SeedIcon />}
                                            value={seed.current}
                                            onChange={event =>
                                                seed.update(event.target.value)
                                            }
                                        />
                                    </Box>
                                }
                            />

                            <Button
                                primary
                                icon={<ShuffleIcon />}
                                label="Shuffle"
                                disabled={isEmpty}
                                onClick={shuffle}
                            />
                        </Box>
                    </Box>
                </Main>
            </Keyboard>
        </Grommet>
    )
}

export default App
