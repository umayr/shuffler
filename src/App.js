import React, { useState } from 'react'
import { Box, Heading, Button, Flex } from 'rebass'
import { Input } from '@rebass/forms'
import shortid from 'shortid'
import { useHistory } from 'react-router-dom'

import useShuffler from './useShuffler'
import useQuery from './useQuery'

import * as base64 from './base64'

const EMPTY = ''

function App() {
  const params = useQuery()
  const history = useHistory()
  const [value, setValue] = useState(EMPTY)
  const [showPreview, setShowPreview] = useState(false)

  const [list, setList, shuffleFn] = useShuffler(params.getAll('item') ?? [])

  const onChange = event => {
    setValue(event.target.value)
  }

  const onSubmit = event => {
    event.preventDefault()

    setList([...list, value])
    setValue(EMPTY)

    params.append('item', value)
    history.push(`/?${params.toString()}`)

    return false
  }

  const onClick = event => {
    event.preventDefault()

    setShowPreview(true)
  }

  const onShuffle = event => {
    event.preventDefault()

    shuffleFn()
  }

  return (
    <Flex
      width={[1, 1, 1 / 2]}
      flexWrap="wrap"
      flexDirection="column"
      p={[4, 4, 5]}
    >
      <Box py={[2, 2, 4]}>
        <form onSubmit={onSubmit}>
          <Input
            id="text"
            name="text"
            type="text"
            value={value}
            onChange={onChange}
            placeholder="Anything goes here"
            autoComplete="off"
          />
        </form>
      </Box>
      {list.length > 0 && (
        <>
          <Box>
            {list.map(item => (
              <Heading
                fontSize={[5, 6, 7]}
                color="primary"
                key={shortid.generate()}
              >
                {item}
              </Heading>
            ))}
          </Box>
          <Box>
            <Flex my={[3, 3, 4]} flexDirection={['column', 'column', 'row']}>
              <Box width={[1, 1, 1 / 2]} mr={[0, 0, 2]} my={[2, 2, 0]}>
                <Button variant="primary" onClick={onShuffle} width={1}>
                  Shuffle
                </Button>
              </Box>
              <Box width={[1, 1, 1 / 2]} mt={[2, 2, 0]}>
                <Button variant="outline" onClick={onClick} width={1}>
                  Share
                </Button>
              </Box>
            </Flex>

            {showPreview && (
              <Input
                id="preview"
                name="preview"
                type="text"
                readOnly
                value={`${window.location.origin}/preview/${base64.to(list)}`}
              />
            )}
          </Box>
        </>
      )}
    </Flex>
  )
}

export default App
