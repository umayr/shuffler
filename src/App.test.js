import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import App from './App'

import { MemoryRouter as Router } from 'react-router-dom'

test('renders app', () => {
  const { container } = render(<App />, { wrapper: Router })

  expect(container).toMatchSnapshot()
})

test('renders app with predefined items in the query params', () => {
  const { container } = render(
    <Router
      initialEntries={[{ pathname: '/', search: 'item=foo&item=bar' }]}
      initialIndex={0}
    >
      <App />
    </Router>,
  )

  expect(container).toHaveTextContent(/foo/)
  expect(container).toHaveTextContent(/bar/)
})

test('populates data from text field', () => {
  const { container, getByPlaceholderText } = render(<App />, {
    wrapper: Router,
  })

  const input = getByPlaceholderText('Anything goes here')
  const form = container.querySelector('form')

  fireEvent.input(input, { target: { value: 'foo' } })
  fireEvent.submit(form)
  expect(container).toHaveTextContent(/foo/)

  fireEvent.input(input, { target: { value: 'bar' } })
  fireEvent.submit(form)
  expect(container).toHaveTextContent(/bar/)

  fireEvent.input(input, { target: { value: 'baz' } })
  fireEvent.submit(form)
  expect(container).toHaveTextContent(/baz/)
})

test('shows preview url', () => {
  const { container, getByPlaceholderText, getByText } = render(
    <App />,
    {
      wrapper: Router,
    },
  )

  const input = getByPlaceholderText('Anything goes here')
  const form = container.querySelector('form')

  fireEvent.input(input, { target: { value: 'foo' } })
  fireEvent.submit(form)
  expect(container).toHaveTextContent(/foo/)

  fireEvent.click(getByText('Share'))

  const preview = container.querySelector('input[name=preview]');

  expect(preview.value).toBe('http://localhost/preview/WyJmb28iXQ==')
})
