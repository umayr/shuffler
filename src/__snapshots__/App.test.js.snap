// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders app 1`] = `
<div>
  <div
    class="css-1fiyid0"
  >
    <div
      class="css-18ljfa2"
    >
      <form>
        <input
          autocomplete="off"
          class="css-4y926w"
          id="text"
          name="text"
          placeholder="Anything goes here"
          type="text"
          value=""
        />
      </form>
    </div>
  </div>
</div>
`;
